src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +672,501 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+              // it is safe since they are not modified by any shard for the
+              // duration of the capture and they are protected from deletion
+              // for the duration of the capture by a gate. Both are used
+              // immediately on the foreign shard to access data local to that
+              // shard. This is meant to help avoiding unintended cross shard
+              // access.
+              return shard_fetch_worker(shard_local_fetch_context{
+                .foreign_read = foreign_read,
+                .bytes_left = octx.bytes_left,
+                .min_bytes = min_fetch_bytes,
+                .deadline = octx.deadline,
+                .requests = std::move(configs),
+                .srv = octx.rctx.server().local(),
+                .mgr = mgr,
+                .as = _as.local(),
+              });
+          });
+
+        fill_fetch_responses(
+          octx,
+          std::move(results.read_results),
+          fetch.responses,
+          fetch.start_time);
+
+        // Remove errored fetches from fetch.requests and fetch.responses.
+        if (!results.errored_partitions.empty()) {
+            std::sort(
+              results.errored_partitions.begin(),
+              results.errored_partitions.end());
+            size_t shift = 0;
+            for (auto i : results.errored_partitions) {
+                fetch.requests.erase(fetch.requests.begin() + (i - shift));
+                fetch.responses.erase(fetch.responses.begin() + (i - shift));
+                shift += 1;
+            }
+        }
+
+        _last_result_size[shard] = results.total_size;
+        _completed_shard_fetches.push_back(std::move(fetch));
+        _has_data.signal();
+    }
+
+    std::optional<ss::future<>> register_hwm_wait(
+      shard_local_fetch_context& ctx,
+      const model::ktp& ktp,
+      model::offset offset,
+      ss::abort_source& as) {
+        auto part = ctx.mgr.get(ktp);
+        // If the partition can't be found then it's since been moved
+        if (!part) {
+            return {};
+        }
+
+        auto consensus = part->raft();
+        if (!consensus) {
+            return {};
+        }
+
+        return {ignore_exceptions(consensus->visible_offset_monitor().wait(
+          offset, model::no_timeout, as))};
+    }
+
+    /**
+     * The shard fetch worker will first attempt to query every partition in
+     * `fetch` for data.
+     *
+     * If there is enough data to meet `min_fetch_bytes` or if
+     * the fetch coordinator has sent an abort then the worker returns.
+     *
+     * Otherwise it'll wait for the last_visible_index to change on one of the
+     * partitions in `fetch` then attempt to gather enough data again.
+     *
+     * This will repeat until `min_fetch_bytes` is met or an abort is received
+     * from the coordinator.
+     */
+    ss::future<worker_result>
+    shard_fetch_worker(shard_local_fetch_context ctx) {
+        auto requests = ctx.requests;
```

Comment:
why make this copy?

Reply:
`ctx.requests` is used later in the function when we have to re-query a partition after it's HWM has increased.

Reply:
OK, this might be an area for future improvement, e.g., sharing the requests in some way if they are not modified, since I guess it can be an expesnive copy for wide reads.